### PR TITLE
Basic text area support and debug log tail

### PR DESCRIPTION
**Pull request by allendav**
Next item for #153 - adds debug log tail and debug on/off toggle

Also tweaks save form to work from store state instead of state mapped to prop

To test:
Navigate to wp-admin > WooCommerce > System Status > WooCommerce Connect
Scroll down to the debug section
See the debug log snippet
Make sure the link to the full log works
Make sure turning logging off and on works

### CODE REVIEW

**Comment by allendav** [comment]
@kellychoffman - ready for styling if you want - will have debug on/off connected shortly

**Comment by kellychoffman** [comment]
Thanks @allendav. Having a look at it now.

**Comment by kellychoffman** [comment]
Styled!

screen shot 2016-06-02 at 1 10 32 pm

**Comment by DanReyLop** [comment]
Everything works (haven't looked at the code yet), except in my browser (Chrome on Windows 10), the logs appear in a single line:
debug_log

The full log shows fine.

File: classes/class-wc-connect-help-provider.php [hunk 1]
```diff
@@ -237,8 +244,72 @@ protected function get_services_items() {
...
+				'saveontoggle' => true
+			);
+
+			// add connect log tail
+			$log_data = $this->get_debug_log_data();
+			if ( count( $log_data->tail ) < 1 ) {
+				$description = '';
+				$log_tail = __( 'Log is empty', 'woocommerce' );
+			} else {
+				$url = admin_url( 'admin.php?page=wc-status&tab=logs&log_file=' . $log_data->key );
```
Comment (jkudish) [inline classes/class-wc-connect-help-provider.php]: please use `add_query_arg`

Comment (allendav) [inline classes/class-wc-connect-help-provider.php]: LOL - my bad. I should have known better :P

**Comment by allendav** [comment]
@DanReyLop : oooh - good catch - ty!

File: classes/class-wc-connect-help-provider.php (hunk 1 above)
Comment (allendav) [inline classes/class-wc-connect-help-provider.php]: Fixed in https://github.com/Automattic/woocommerce-connect-client/pull/376/commits/5fce8ed4e0a16b1e32d806a6b28b329e94a94069

**Comment by allendav** [comment]
@DanReyLop fixed wrapping problem in https://github.com/Automattic/woocommerce-connect-client/pull/376/commits/6b1b211a82455e4a18c72af0f19ecf6351b6b97a

**Comment by allendav** [comment]
@jkudish @DanReyLop - let me know if this is good to go - ty!

File: classes/class-wc-connect-help-provider.php [hunk 2]
```diff
@@ -237,8 +244,72 @@ protected function get_services_items() {
...
+		/**
+		 * Gets the last 10 lines from the WooCommerce Connect log, if it exists
+		 */
+		protected function get_debug_log_data() {
+			$data = new stdClass;
+			$data->key = '';
+			$data->file = '';
+			$data->tail = array();
+
+			if ( method_exists( 'WC_Admin_Status', 'scan_log_files' ) ) {
```
Comment (DanReyLop) [inline classes/class-wc-connect-help-provider.php]: Why wouldn't this method exist?

File: classes/class-wc-connect-help-provider.php [hunk 3]
```diff
@@ -339,7 +417,23 @@ protected function get_form_schema() {
...
+						);
+					}
+
+					if ( 'boolean' === $fieldsetitem->type ) {
+						$form_properties[ $fieldsetitem->key ] = array(
+							'title' => $fieldsetitem->title,
+							'type' => 'boolean',
+							'text' => property_exists( $fieldsetitem, 'text' ) ? $fieldsetitem->text : '',
+							'description' => property_exists( $fieldsetitem, 'description' ) ? $fieldsetitem->description : '',
+							'saveontoggle' => property_exists( $fieldsetitem, 'saveontoggle' ) ? $fieldsetitem->saveontoggle : '',
```
Comment (DanReyLop) [inline classes/class-wc-connect-help-provider.php]: The default value for ```saveontoggle``` should be ```FALSE``` instead of an empty string. It doesn't really matter because Javascript will check for truthy values but still.

File: classes/class-wc-connect-help-provider.php [hunk 4]
```diff
@@ -416,13 +517,15 @@ public function page() {
 			$form_schema = new stdClass();
 			$form_schema->properties = new stdClass();
 
+			$path = "/wc/v1/connect/self-help";
```
Comment (DanReyLop) [inline classes/class-wc-connect-help-provider.php]: Unused variable

File: classes/class-wc-connect-logger.php [hunk 5]
```diff
@@ -2,16 +2,21 @@
...
 		 * @var WC_Logger
 		 */
 		private $logger;
 
+		private $is_logging_enabled = false;
+
 		public function __construct( WC_Logger $logger ) {
 
 			$this->logger = $logger;
+			$this->is_logging_enabled = ( 1 == get_option( WOOCOMMERCE_CONNECT_DEBUG_LOGGING_ENABLED_OPTION, 0 ) );
```
Comment (DanReyLop) [inline classes/class-wc-connect-logger.php]: If you are updating the option with true / false values, then comparing against ```1``` is redundant. Please don't rely on PHP ```==``` comparison, is the most broken feature of the language :(

File: client/components/toggle/style.scss [hunk 6]
```diff
@@ -0,0 +1,9 @@
+.toggle__text {
+	color: $gray;
+	font-size: 11px;
+	line-height: 16px;
+	margin-left: 8px;
+	vertical-align: middle;
+	text-transform: uppercase;
+	cursor: pointer;
```
Comment (DanReyLop) [inline client/components/toggle/style.scss]: The toggle text is not actually clickable, so it's better to actually make it clickable or remove the "hand" pointer.

File: package.json [hunk 7]
```diff
@@ -78,6 +78,7 @@
     "react-dom": "0.14.7",
     "react-redux": "^4.4.1",
     "redux": "^3.3.1",
+    "redux-thunk": "^2.1.0",
```
Comment (DanReyLop) [inline package.json]: Why do we need this? You aren't using ```thunk```in the changes introduces in this PR

File: classes/class-wc-connect-help-provider.php (hunk 2 above)
Comment (allendav) [inline classes/class-wc-connect-help-provider.php]: Presently it can't (if WooCommerce has not been activated, you can't even get to this page), but I plan on splitting out the problem detection from the problem presentation in an upcoming PR, so it will make more sense then. (The split will be needed to enable things like wp-admin notices and this debug page to have a single source of truth.)

The funny thing is, the window will be short - 'cause once we are part of WooCommerce this will be moot again, but a window will be there for a little bit, so I'd like to leave this in.

File: package.json (hunk 7 above)
Comment (allendav) [inline package.json]: Good catch. Leftover from previous abandoned work on this PR. Will remove.

File: classes/class-wc-connect-logger.php (hunk 5 above)
Comment (allendav) [inline classes/class-wc-connect-logger.php]: I was getting odd behavior without this before. Will retest and simplify.

**Comment by jkudish** [comment]
is it expected that it says enabled but is greyed out?

image

File: classes/class-wc-connect-help-provider.php [hunk 8]
```diff
@@ -237,8 +244,79 @@ protected function get_services_items() {
...
+					wp_kses(
+						__( 'Last %d entries <a href="%s">Show full log</a>', 'woocommerce' ),
+						array(  'a' => array( 'href' => array() ) ) ),
+					count( $log_data->tail ),
+					esc_url( $url )
+				);
+				$log_tail = implode( $log_data->tail, '' );
+			}
+
+			$debug_items[] =	(object) array(
```
Comment (jkudish) [inline classes/class-wc-connect-help-provider.php]: space after `=`

File: classes/class-wc-connect-help-provider.php [hunk 9]
```diff
@@ -237,8 +244,79 @@ protected function get_services_items() {
...
+						'tab' => 'logs',
+						'log_file' => $log_data->key
+					),
+					admin_url( 'admin.php' )
+				);
+				$description = sprintf(
+					wp_kses(
+						__( 'Last %d entries <a href="%s">Show full log</a>', 'woocommerce' ),
+						array(  'a' => array( 'href' => array() ) ) ),
+					count( $log_data->tail ),
```
Comment (jkudish) [inline classes/class-wc-connect-help-provider.php]: this count is repeated twice so could be DRYed

File: classes/class-wc-connect-logger.php [hunk 10]
```diff
@@ -2,16 +2,21 @@
 
 if ( ! class_exists( 'WC_Connect_Logger' ) ) {
 
+	define( 'WOOCOMMERCE_CONNECT_DEBUG_LOGGING_ENABLED_OPTION', 'wcc_debug_logging_enabled' );
```
Comment (jkudish) [inline classes/class-wc-connect-logger.php]: why bother with a constant?

File: classes/class-wc-connect-logger.php [hunk 11]
```diff
@@ -38,25 +43,42 @@ protected function format_message( $message, $context = '' ) {
...
 		}
 
+		public function enable_logging() {
+			update_option( WOOCOMMERCE_CONNECT_DEBUG_LOGGING_ENABLED_OPTION, true );
+			$this->is_logging_enabled = true;
+			$this->log( "Logging enabled" );
+		}
+
+		public function disable_logging() {
+			$this->log( "Logging disabled" );
```
Comment (jkudish) [inline classes/class-wc-connect-logger.php]: very very minor but i had to double take here because the order of operations here is different then for enable

File: client/components/text-area/index.js [hunk 12]
```diff
@@ -0,0 +1,59 @@
+import React, { PropTypes } from 'react';
+import FormFieldset from 'components/forms/form-fieldset';
+import FormLabel from 'components/forms/form-label';
+import FormTextarea from 'components/forms/form-textarea';
+import FormSettingExplanation from 'components/forms/form-setting-explanation';
+import FormInputValidation from 'components/forms/form-input-validation';
+import { sanitize } from 'dompurify';
+
+const renderFieldDescription = ( description ) => {
```
Comment (jkudish) [inline client/components/text-area/index.js]: this is repeated across several components now, so prob an opportunity to DRY (can be a in a separate PR)

File: client/components/text-area/index.js [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+	return (
+		description ? <FormSettingExplanation dangerouslySetInnerHTML={ { __html: sanitize( description, { ADD_ATTR: ['target'] } ) } } /> : null
+	);
+};
+
+const renderFieldError = ( validationHint ) => {
```
Comment (jkudish) [inline client/components/text-area/index.js]: same here... this is repeated across several components now, so prob an opportunity to DRY (can be a in a separate PR)

File: client/components/text-area/style.scss [hunk 14]
```diff
@@ -0,0 +1,10 @@
+.form-textarea {
+	font-size: 12px;
+	height: 195px;
+	resize: none;
+	white-space: pre;
+
+	&#wcc_debug_log_tail {
+		font-family: Consolas, "Andale Mono WT", "Andale Mono", "Lucida Console", Monaco, "Courier New", Courier, monospace;
```
Comment (jkudish) [inline client/components/text-area/style.scss]: why so many?

File: client/components/toggle/index.js [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+import FormSettingExplanation from 'components/forms/form-setting-explanation';
+import { sanitize } from 'dompurify';
+
+const renderToggleText = ( text ) => {
+	return (
+		text ? <span className="toggle__text">{ text }</span> : null
+	);
+};
+
+const renderFieldDescription = ( description ) => {
```
Comment (jkudish) [inline client/components/toggle/index.js]: same here... this is repeated across several components now, so prob an opportunity to DRY (can be a in a separate PR)

File: client/components/text-area/style.scss (hunk 14 above)
Comment (allendav) [inline client/components/text-area/style.scss]: ^^ @kellychoffman ?

File: client/components/text-area/index.js (hunk 13 above)
Comment (allendav) [inline client/components/text-area/index.js]: Agreed

File: classes/class-wc-connect-logger.php (hunk 11 above)
Comment (allendav) [inline classes/class-wc-connect-logger.php]: Yeah, its funny, 'cause I wanted it to actually log the state change :)

File: classes/class-wc-connect-logger.php (hunk 10 above)
Comment (allendav) [inline classes/class-wc-connect-logger.php]: Because the constant is used in multiple places - didn't want to risk a change not hitting all the instances of the option name

**Comment by allendav** [comment]
> is it expected that (when enabled) it says enabled but is greyed out?

^^ @kellychoffman

File: classes/class-wc-connect-logger.php (hunk 5 above)
Comment (allendav) [inline classes/class-wc-connect-logger.php]: OK, I tested this and remembered why I needed this: update_option with true results in a string 1 being returned by get_option.

Similarly, update_option with false results in an empty string being returned by get_option

Here's an article that describes the stringification behavior a bit - https://www.storelocatorplus.com/wordpress-boolean-options-and-javascript-handle-with-care/

Net net - I'm going to change that get_option to a strict === comparison for string '1' and make the default a string '0'

Comment (jkudish) [inline classes/class-wc-connect-logger.php]: Why not just `$this->is_logging_enabled = (bool) get_option( WOOCOMMERCE_CONNECT_DEBUG_LOGGING_ENABLED_OPTION );` ?

File: classes/class-wc-rest-connect-self-help-controller.php [hunk 16]
```diff
@@ -0,0 +1,88 @@
+<?php
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+if ( ! defined( 'WOOCOMMERCE_CONNECT_MAX_SELF_HELP_DEPTH' ) ) {
+	define( 'WOOCOMMERCE_CONNECT_MAX_SELF_HELP_DEPTH', 32 );
```
Comment (jkudish) [inline classes/class-wc-rest-connect-self-help-controller.php]: why not just re-use https://github.com/Automattic/woocommerce-connect-client/blob/9ca816e35286eaef4727fe9b64b6b6a3a61dc380/classes/class-wc-rest-connect-services-controller.php#L8 ?

File: client/components/text-area/index.js (hunk 12 above)
Comment (allendav) [inline client/components/text-area/index.js]: Agreed - https://github.com/Automattic/woocommerce-connect-client/issues/393

File: client/components/toggle/index.js (hunk 15 above)
Comment (allendav) [inline client/components/toggle/index.js]: Agreed - https://github.com/Automattic/woocommerce-connect-client/issues/393

File: classes/class-wc-rest-connect-self-help-controller.php (hunk 16 above)
Comment (allendav) [inline classes/class-wc-rest-connect-self-help-controller.php]: Sure - I'll create a single constant for all these json_decode limiters and use it throughout WCC

File: classes/class-wc-connect-help-provider.php (hunk 8 above)
Comment (allendav) [inline classes/class-wc-connect-help-provider.php]: Fixed in https://github.com/Automattic/woocommerce-connect-client/pull/376/commits/77e6c33e877d6931f17e4958e855171bc1cf814b

File: package.json (hunk 7 above)
Comment (allendav) [inline package.json]: Fixed in https://github.com/Automattic/woocommerce-connect-client/pull/376/commits/77e6c33e877d6931f17e4958e855171bc1cf814b

File: client/components/toggle/style.scss (hunk 6 above)
Comment (allendav) [inline client/components/toggle/style.scss]: Removed pointer in https://github.com/Automattic/woocommerce-connect-client/pull/376/commits/77e6c33e877d6931f17e4958e855171bc1cf814b

File: classes/class-wc-connect-logger.php (hunk 5 above)
Comment (allendav) [inline classes/class-wc-connect-logger.php]: Fixed in https://github.com/Automattic/woocommerce-connect-client/pull/376/commits/77e6c33e877d6931f17e4958e855171bc1cf814b

File: classes/class-wc-connect-help-provider.php (hunk 4 above)
Comment (allendav) [inline classes/class-wc-connect-help-provider.php]: Fixed in https://github.com/Automattic/woocommerce-connect-client/pull/376/commits/77e6c33e877d6931f17e4958e855171bc1cf814b

File: classes/class-wc-connect-help-provider.php (hunk 3 above)
Comment (allendav) [inline classes/class-wc-connect-help-provider.php]: Fixed in https://github.com/Automattic/woocommerce-connect-client/pull/376/commits/77e6c33e877d6931f17e4958e855171bc1cf814b

File: classes/class-wc-connect-help-provider.php (hunk 9 above)
Comment (allendav) [inline classes/class-wc-connect-help-provider.php]: Fixed in https://github.com/Automattic/woocommerce-connect-client/pull/376/commits/77e6c33e877d6931f17e4958e855171bc1cf814b

File: classes/class-wc-rest-connect-self-help-controller.php (hunk 16 above)
Comment (allendav) [inline classes/class-wc-rest-connect-self-help-controller.php]: Fixed in https://github.com/Automattic/woocommerce-connect-client/pull/376/commits/77e6c33e877d6931f17e4958e855171bc1cf814b

**Comment by allendav** [comment]
@jkudish @DanReyLop : while we wait for @kellychoffman 's feedback on the two items above, let me know if there are further changes needed for this

File: client/components/text-area/style.scss (hunk 14 above)
Comment (kellychoffman) [inline client/components/text-area/style.scss]: Why so many fonts? Because you never know what fonts people are going to have. I even shaved a few off.

**Comment by kellychoffman** [comment]
> is it expected that (when enabled) it says enabled but is greyed out?

Not at all :) Thanks for catching. The text should say 'Disabled' when its disabled.

**Comment by allendav** [comment]
TODO - change text on toggle from "Enabled" to "Disabled" when the toggle is set to off (and vice versa) per @kellychoffman

**Comment by allendav** [comment]
@jkudish @DanReyLop @kellychoffman - ok, with https://github.com/Automattic/woocommerce-connect-client/pull/376/commits/44a4a4d74df4efb5f6c21391677845fdeefe847b it now has different text for different states - please let me know if this is good to go - thx!

**Comment by DanReyLop** [comment]
:shipit:
